### PR TITLE
docs: add request size guide

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
Thanks for your comments!

### CODE REVIEW

File: chapters/request_size.rst [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+``request_size`` is the parameter that we could set for the Flow.
+It defines the number of data sent into parallel ``Peas``.
+
+.. contents:: Table of Contents
+    :depth: 2
+
+Request Size
+------------
+Jina defines the ``request_size`` as the parameter at client side. By setting the ``request_size`` in Flow's API
+we could define the number of data sent to each ``Peas``.
```
Comment (JoanFM) [inline chapters/request_size.rst]: we could define the `number` of Documents contained in every `Request`.

File: chapters/request_size.rst [hunk 2]
```diff
@@ -0,0 +1,133 @@
+=================
+Request Size
+=================
+
+``request_size`` is the parameter that we could set for the Flow.
+It defines the number of data sent into parallel ``Peas``.
```
Comment (JoanFM) [inline chapters/request_size.rst]: This is not accurate, it defines the number of `Document` a single `Request` will contain

Comment (JoanFM) [reply chapters/request_size.rst]: so every Pea will receive requests containing this amount of documents

File: chapters/request_size.rst [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+------------
+Batch size, commonly used in machine learning, often refers to the number of data that we feed into
+the model in one iteration.
+
+In Jina, ``batch_size`` is used at ``Driver`` and ``Executor`` level. Where
+we align the usage of machine learning terms for ``batch_size``.
+
+Before you start
+------------------
+Make sure you intall Numpy and latest version of Jina on your local machine.
```
Comment (JoanFM) [inline chapters/request_size.rst]: Latest version of Jina already installs numpy as it is a core dependency

File: chapters/request_size.rst [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        def craft(self, id, *args, **kwargs):
+            return {'id': id}
+
+
+Next we define the data generator.
+
+.. highlight:: python
+.. code-block:: python
+
+    def random_docs(num_docs, chunks_per_doc=5, embed_dim=10, jitter=1, start_id=0, embedding=True) -> Iterator['Document']:
```
Comment (JoanFM) [inline chapters/request_size.rst]: This is way too noisy to explain the concept

File: chapters/request_size.rst [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        main()
+
+
+Choose different request size
+--------------------
+Different setting of ``request_size`` may influence the running performance.
+A higher value means large size data will be fed into the ``Pea`` and will demand more memory.
+A lower value will decrease the cost of memory but may increase the running time.
+
+An simple extension of the above example generate a box plot showing the relationship between ``request_size`` and running time
```
Comment (JoanFM) [inline chapters/request_size.rst]: do we have this extension?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: chapters/request_size.rst (hunk 4 above)
Comment (Yongxuanzhang) [reply chapters/request_size.rst]: How about just using numpy.random?

File: chapters/request_size.rst (hunk 3 above)
Comment (Yongxuanzhang) [reply chapters/request_size.rst]: ✅

File: chapters/request_size.rst (hunk 4 above)
Comment (JoanFM) [reply chapters/request_size.rst]: yes anything where the request size concept shines with easy Documents

File: chapters/request_size.rst (hunk 5 above)
Comment (Yongxuanzhang) [reply chapters/request_size.rst]: No, we don't. That just means we extend the code example above by setting different request size values to help users get more insights on the influence of this parameter. Not sure if this is a good 😄

File: chapters/request_size.rst (hunk 1 above)
Comment (Yongxuanzhang) [reply chapters/request_size.rst]: ✔️

File: chapters/request_size.rst (hunk 2 above)
Comment (Yongxuanzhang) [reply chapters/request_size.rst]: ✅

File: chapters/request_size.rst [hunk 6]
```diff
@@ -3,15 +3,15 @@ Request Size
 =================
 
 ``request_size`` is the parameter that we could set for the Flow.
-It defines the number of data sent into parallel ``Peas``.
+It defines the number of ``Document`` a single ``Request`` will contain, so every ``Pea`` will receive requests containing this amount of documents.
```
Comment (JoanFM) [inline chapters/request_size.rst]: I would remove the word every

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: chapters/request_size.rst (hunk 6 above)
Comment (Yongxuanzhang) [reply chapters/request_size.rst]: Let's say we have 100 Documents, 10 crafters and request_size=10, this means that we will generate 100/10=10 requests each contains 10 docs(request_size) and will be sent to crafters via load balancing or round robin? Is that correct?

File: chapters/request_size.rst (hunk 4 above)
Comment (Yongxuanzhang) [reply chapters/request_size.rst]: ✅

File: chapters/request_size.rst (hunk 6 above)
Comment (Yongxuanzhang) [reply chapters/request_size.rst]: Like the free Pea will receive the request according to our SchedulerType

File: chapters/request_size/request_size.rst [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+------------
+Jina defines the ``request_size`` as the parameter at client side. By setting the ``request_size`` in Flow's API
+we could define the number of ``Documents`` contained in every ``Request``.
+
+Batch Size
+------------
+Batch size, commonly used in machine learning, often refers to the number of data that we feed into
+the model in one iteration.
+
+In Jina, ``batch_size`` is used at ``Driver`` and ``Executor`` level. Where
```
Comment (JoanFM) [inline chapters/request_size/request_size.rst]: In Jina batch size is used by drivers and Encoders to guarantee that the Executor processes data in pieces of a specific size

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: chapters/request_size/request_size.rst (hunk 7 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: Only Encoders use batch size?

File: chapters/request_size.rst (hunk 6 above)
Comment (JoanFM) [reply chapters/request_size.rst]: yes, so if polling is not all, not every Pea will see all these Documents

File: chapters/request_size/request_size.rst (hunk 7 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: Done

Comment (JoanFM) [reply chapters/request_size/request_size.rst]: no sorry Executors, Indexers also do

Comment (JoanFM) [reply chapters/request_size/request_size.rst]: so better say Executors sorry

Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: Got it. ✅

File: chapters/request_size/request_size.rst [hunk 8]
```diff
@@ -0,0 +1,99 @@
+=================
+Request Size
```
Comment (FionnD) [inline chapters/request_size/request_size.rst]: A slightly more descriptive title would be great! 

Something like: **Understanding the Request Size parameter**? 

What do you think?

File: chapters/request_size/request_size.rst [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+=================
+
+``request_size`` is the parameter that we could set for the Flow.
```
Comment (FionnD) [inline chapters/request_size/request_size.rst]: I think we can shorten this to:

In some use cases, you may wish to vary the number of Documents a single Request will receive. You can achieve this by adjusting the request_size parameter when setting the Flow. This guide covers the different configuration options Jina offers.

File: chapters/request_size/request_size.rst [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+so ``Peas`` will receive requests containing this amount of ``Documents`` according to different
+scheduling policies.
+
+.. contents:: Table of Contents
+    :depth: 2
+
+Request Size
+------------
+Jina defines the ``request_size`` as the parameter at client side. By setting the ``request_size`` in Flow's API
+we could define the number of ``Documents`` contained in every ``Request``.
```
Comment (FionnD) [inline chapters/request_size/request_size.rst]: Jina defines the ``request_size`` as **a parameter on the** client side. By **adjusting** the ``request_size`` in **the** Flow's API **you** can define the number of ``Documents`` contained in every ``Request``.

File: chapters/request_size/request_size.rst [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+Batch Size
+------------
+Batch size, commonly used in machine learning, often refers to the number of data that we feed into
```
Comment (FionnD) [inline chapters/request_size/request_size.rst]: number of data**points**

File: chapters/request_size/request_size.rst [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+.. highlight:: bash
+.. code-block:: bash
+
+    pip install -U jina
+
+Implementation
+--------------------
+In order to have a better understanding of the influence of ``request_size`` and how it is used, let's take the following
+codes as example.
```
Comment (FionnD) [inline chapters/request_size/request_size.rst]: let's take the following **code snippets as examples**

File: chapters/request_size/request_size.rst [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+        end_time = time.time()
+        seconds_elapsed = end_time - start_time
+        print(seconds_elapsed)
+
+
+    if __name__ == '__main__':
+        main()
+
+
+Choose different request size
```
Comment (FionnD) [inline chapters/request_size/request_size.rst]: **Choosing**

File: chapters/request_size/request_size.rst [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+------------------------------
+Different setting of ``request_size`` may influence the running performance.
```
Comment (FionnD) [inline chapters/request_size/request_size.rst]: Different setting**s** of

File: chapters/request_size/request_size.rst [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+A higher value means large size data will be fed into the ``Pea`` and will demand more memory.
```
Comment (FionnD) [inline chapters/request_size/request_size.rst]: A higher value means large **Documents** will be fed into the ``Pea`` and will demand more memory.

File: chapters/request_size/request_size.rst [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+A lower value will decrease the cost of memory but may increase the running time since we need to send more ``requests``.
+
+An simple extension of the above example generate a box plot showing the relationship between ``request_size`` and running time
+when we have 100 number of ``Documents`` to be indexed.
+which may help you to get more insights on choosing the ``request_size``.
```
Comment (FionnD) [inline chapters/request_size/request_size.rst]: A simple extension of the above example generates a box plot showing the relationship between ``request_size`` and running time when we have 100 ``Documents`` to be indexed. This may help you to get more insights on choosing the ``request_size``.

**Review by FionnD — CHANGES_REQUESTED** [review]
A really like this and learned a lot! The graph is very cool.

Some grammar suggestions that might make it flow better

File: chapters/request_size/request_size.rst (hunk 9 above)
Comment (cristianmtr) [inline chapters/request_size/request_size.rst]: Don't break the sentences within a paragraph. Remove the `\n`

File: chapters/request_size/request_size.rst [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+.. highlight:: python
+.. code-block:: python
+
+    class SimpleCrafter(BaseCrafter):
+        def craft(self, id, *args, **kwargs):
+            return {'id': id}
+
+
+For this example, we will index 100 documents and use 10 parallel ``Crafters``. The ``request_size``
+is set as 20. So the 100 ``Documents`` will be divided into 5 parts and each ``Request`` contains 20 ``Documents``.
```
Comment (cristianmtr) [inline chapters/request_size/request_size.rst]:
```suggestion
is set to 20. So the 100 ``Documents`` will be divided into 5 parts and each ``Request`` contains 20 ``Documents``.
```

File: chapters/request_size/request_size.rst (hunk 14 above)
Comment (cristianmtr) [inline chapters/request_size/request_size.rst]: Again, don't break sentences within a paragraph

**Review by cristianmtr — DISMISSED** [review]
see comments

File: chapters/request_size/request_size.rst (hunk 8 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: Is it ok to use a question-like title? 😂  I'm ok with it but not sure if it suits the whole doc style

File: chapters/request_size/request_size.rst (hunk 9 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: ✅

File: chapters/request_size/request_size.rst (hunk 10 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: 👍

File: chapters/request_size/request_size.rst (hunk 11 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: 👍

File: chapters/request_size/request_size.rst (hunk 13 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: 👍

File: chapters/request_size/request_size.rst (hunk 14 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: 👍

File: chapters/request_size/request_size.rst (hunk 16 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: 👍

**Comment by Yongxuanzhang** [comment]
> A really like this and learned a lot! The graph is very cool.
> 
> Some grammar suggestions that might make it flow better

Thank you so much @FionnD! Almost help me to rewrite this section 😂 .

File: chapters/request_size/request_size.rst (hunk 17 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: 👍

File: chapters/request_size/request_size.rst (hunk 14 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: Are there any reasons we don't break them? They are actually in one paragraph when deployed, adding one more blank line will change the line.

File: chapters/request_size/request_size.rst (hunk 8 above)
Comment (FionnD) [reply chapters/request_size/request_size.rst]: Ah I didn't mean to add the **?**. 

I think: 
Understanding the Request Size parameter
is good

File: chapters/request_size/request_size.rst [hunk 18]
```diff
@@ -0,0 +1,87 @@
+=============================================
+ Understanding the Request Size parameter?
```
Comment (FionnD) [inline chapters/request_size/request_size.rst]: As noted in the other comment, just remove the ?. 

That is my fault, my previous comment was unclear

File: chapters/request_size/request_size.rst [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+=============================================
+
+In some use cases, you may wish to vary the number of ``Documents`` a single Request will receive. You can achieve this by adjusting the ``request_size`` parameter when setting the Flow. This guide covers the different configuration options Jina offers.
+
+.. contents:: Table of Contents
+    :depth: 2
+
+Request Size
+------------
+Jina defines the ``request_size`` as the parameter on client side. By adjusting the ``request_size`` in the Flow's API you could define the number of ``Documents`` contained in every ``Request``.
```
Comment (FionnD) [inline chapters/request_size/request_size.rst]: on **the** client side.

Comment (FionnD) [inline chapters/request_size/request_size.rst]: Flow's API you **can** define

**Review by FionnD — CHANGES_REQUESTED** [review]
One comment, and two small fixes which I had missed in the first round. 

You can merge this once fixed :)

File: chapters/request_size/request_size.rst (hunk 18 above)
Comment (bwanglzu) [reply chapters/request_size/request_size.rst]: same question where question mark here?

File: chapters/request_size/request_size.rst [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+
+Batch Size
+------------
+Batch size, commonly used in machine learning, often refers to the number of datapoints that we feed into the model in one iteration.
+
+In Jina ``batch_size`` is used by ``Drivers`` and ``Executors`` to guarantee that the ``Executor`` processes data in pieces of a specific size.
+
```
Comment (bwanglzu) [inline chapters/request_size/request_size.rst:19]: you can refer to a glossary using :term:`Driver`, checkout glossary [here](https://docs.jina.ai/chapters/glossary/glossary.html). And [how](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/glossary.html) to use glossary.

**Review by bwanglzu — DISMISSED** [review]
good one!

File: chapters/request_size/request_size.rst (hunk 18 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: Oh my bad sorry about that

Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: Oh my bad sorry about that

File: chapters/request_size/request_size.rst (hunk 19 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: ✅

Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: ✅

File: chapters/request_size/request_size.rst [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+--------------------
+In order to have a better understanding of the influence of ``request_size`` and how it is used, let's take the following code snippets as examples.
+
+We first import the necessary modules.
+
+.. highlight:: python
+.. code-block:: python
+
+    import time
+    from typing import Iterator
```
Comment (JoanFM) [inline chapters/request_size/request_size.rst]: Iterator is not needed here

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: chapters/request_size/request_size.rst (hunk 21 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: ✅

File: chapters/request_size/request_size.rst (hunk 20 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst:19]: 👍

**Review by FionnD — APPROVED** [review]

File: chapters/request_size/request_size.rst (hunk 14 above)
Comment (Yongxuanzhang) [reply chapters/request_size/request_size.rst]: ✅

**Review by JoanFM — APPROVED** [review]

**Review by bwanglzu — APPROVED** [review]
LGTM!
